pkg/chains/formats/intotoite6/pipelinerun/pipelinerun.go

```diff
@@ -0,0 +1,219 @@
+		if tr == nil || tr.Status.CompletionTime == nil {
+			logger.Infof("taskrun status not found for task %s", t.Name)
+			continue
+		}
+		steps := []util.StepAttestation{}
+		for i, stepState := range tr.Status.Steps {
+			step := tr.Status.TaskSpec.Steps[i]
+			steps = append(steps, util.AttestStep(&step, &stepState))
+		}
+		after := t.RunAfter
+
+		// Establish task order by retrieving all task's referenced
+		// in the "when" and "params" fields
+		refs := v1beta1.PipelineTaskResultRefs(&t)
+		for _, ref := range refs {
+
+			// Ensure task doesn't already exist in after
+			found := false
+			for _, at := range after {
+				if at == ref.PipelineTask {
+					found = true
+				}
+			}
+			if !found {
+				after = append(after, ref.PipelineTask)
+			}
+		}
+
+		// tr is a finally task without an explicit runAfter value. It must have executed
+		// after the last non-finally task, if any non-finally tasks were executed.
+		if len(after) == 0 && i >= len(pSpec.Tasks) && last != "" {
+			after = append(after, last)
+		}
+		params := tr.Spec.Params
+		var paramSpecs []v1beta1.ParamSpec
+		if tr.Status.TaskSpec != nil {
+			paramSpecs = tr.Status.TaskSpec.Params
+		} else {
+			paramSpecs = []v1beta1.ParamSpec{}
+		}
+		task := TaskAttestation{
+			Name:       t.Name,
+			After:      after,
+			StartedOn:  tr.Status.StartTime.Time,
+			FinishedOn: tr.Status.CompletionTime.Time,
+			Status:     getStatus(tr.Status.Conditions),
+			Steps:      steps,
+			Invocation: util.AttestInvocation(params, paramSpecs, logger),
+			Results:    tr.Status.TaskRunResults,
+		}
+
+		if t.TaskRef != nil {
+			task.Ref = *t.TaskRef
+		}
+
+		tasks = append(tasks, task)
+		if i < len(pSpec.Tasks) {
+			last = task.Name
+		}
+	}
+	return BuildConfig{Tasks: tasks}
+}
+
+func metadata(pro *objects.PipelineRunObject) *slsa.ProvenanceMetadata {
+	m := &slsa.ProvenanceMetadata{}
+	if pro.Status.StartTime != nil {
+		m.BuildStartedOn = &pro.Status.StartTime.Time
+	}
+	if pro.Status.CompletionTime != nil {
+		m.BuildFinishedOn = &pro.Status.CompletionTime.Time
+	}
+	for label, value := range pro.Labels {
+		if label == util.ChainsReproducibleAnnotation && value == "true" {
+			m.Reproducible = true
+		}
+	}
+	return m
+}
+
+// add any Git specification to materials
+func materials(pro *objects.PipelineRunObject) []slsa.ProvenanceMaterial {
+	var mats []slsa.ProvenanceMaterial
+	var commit, url string
+	// search spec.params
+	for _, p := range pro.Spec.Params {
+		if p.Name == util.CommitParam {
+			commit = p.Value.StringVal
+			continue
+		}
+		if p.Name == util.UrlParam {
+			url = p.Value.StringVal
+		}
+	}
+
+	// search status.PipelineSpec.params
+	if pro.Status.PipelineSpec != nil {
+		for _, p := range pro.Status.PipelineSpec.Params {
+			if p.Default == nil {
+				continue
+			}
+			if p.Name == util.CommitParam {
+				commit = p.Default.StringVal
+				continue
+			}
+			if p.Name == util.UrlParam {
+				url = p.Default.StringVal
+			}
+		}
+	}
```

Comment:
I am a bit skeptical about the ordering here. Shouldn't we retrieve the values from `PipelineSpec.Params[x].Default` first and then retrieve from `PipelineRun.Spec.Params[x].Value` so that we are assured that we always get the latest value from pipelinerun level? 

This problem exists for the materials in the taskrun level - task spec level default param value will overwrite the  param value provided from run level.

wdyt?

Reply:
Yes, if a run value was provided it should choose that over the default value. Have you been able to reproduce that behavior? If so it should be fixed. 

Let me know if you want to submit a PR to resolve these, otherwise I'll create a work item to to track them.

Reply:
It should be fixed for TaskRun attestations in the main branch: https://github.com/tektoncd/chains/pull/449
We just need to apply the same fix for PipelineRun attestations as well (and ensure this PR doesn't revert the fix on TaskRuns)

Reply:
#449 fixed the problem in `invocation.parameters`. What I commented on is `materials` section.

https://github.com/tektoncd/chains/pull/527 should fix `materials` section for the taskrun level. We just need to make sure it's correct for pipeline run level attestation in this PR.